feat(registry): support selecting artifacts from OCI Image Index

lexfrei opened the pull request:
**What this PR does / why we need it**:

Implements OCI 1.1 artifact support for Helm charts (HIP: helm/community#424):

1. **Select a chart from an Image Index.** When a reference resolves to an Image Index with multiple manifests (a container image alongside a chart, or several charts), Helm selects the chart by descriptor `artifactType`, disambiguating multiple charts by the requested name then version, and falling back to the chart config for legacy indexes without annotations. An unresolvable choice is an error, never a silent guess.
2. **Set `artifactType` on push.** `helm push` sets the manifest `artifactType` and chart name/version annotations so charts are identifiable at the Index level without fetching each manifest.
3. **`--subject` for the Referrers API.** `helm push --subject <digest>` associates the chart with an existing artifact in the same repository; the digest is validated and resolved against the target repository to a full descriptor.

**Why this is needed:**

Since the index-pull regression was fixed (#31776), Helm traverses an Image Index but selects the chart by a hardcoded media-type scan, last-match-wins — order-dependent and able to return a mislabeled chart for multi-artifact/multi-chart indexes. This makes selection deterministic and lets a chart and its image be co-versioned under one reference.

**Usage:**

```bash
helm pull oci://registry/repo/chart --version 1.0.0           # selects the chart from an index
helm push chart.tgz oci://registry/repo --subject sha256:...  # referrers association
```

**Notes for reviewers:**

- Backward compatible: single-manifest charts are unaffected; a single chart in an index is selected regardless of name.
- `artifactType` is now set on every chart push (per OCI 1.1), which changes the manifest envelope digest versus older Helm for identical chart content; pre-1.1 registries should tolerate the optional field.
- `--subject` is per-repository; cross-repo subjects are out of scope (HIP Open Issues).

Closes #31582

**HIP**: helm/community#424

```release-note
`helm pull` now selects a chart from an OCI Image Index by artifactType (disambiguating multiple charts by name and version), and `helm push` sets the chart artifactType and adds a `--subject` flag to associate a chart with an existing same-repository artifact via the OCI Referrers API.
```

**If applicable**:

- [x] this PR contains user facing changes
- [x] this PR contains unit tests
- [x] this PR has been tested for backwards compatibility

atombrella commented on pkg/registry/constants.go:
```diff
@@ -34,4 +34,8 @@ const (
+
+	// ChartArtifactType is the artifact type for Helm charts in OCI v1.1+ Image Index.
+	// This is used to identify Helm chart manifests within a multi-artifact index.
+	ChartArtifactType = "application/vnd.cncf.helm.config.v1+json"
```
This is already defined in line 27.

atombrella commented on pkg/registry/generic.go:
```diff
@@ -83,6 +89,67 @@ func NewGenericClient(client *Client) *GenericClient {
+
+	return ocispec.Descriptor{}, fmt.Errorf(
+		"no manifest with artifactType %q found in image index; available types: %v",
+		artifactType, availableTypes)
```
Not sure if it's configurable with a linter, but I'd prefer `)` on a separate line. I think that's also more GoLang'ish style to end the last parameter with a  comma `,`

lexfrei committed:
feat(registry): OCI 1.1 artifact support for Helm charts

Add OCI 1.1 artifact support for charts (HIP: helm/community#424):

- Select a chart from an OCI Image Index by descriptor artifactType,
  disambiguating multiple charts by the requested name then version, and
  falling back to the chart config for legacy indexes without
  annotations. An unresolvable choice errors with the candidate list
  instead of returning a position-dependent, mislabeled match.
- Set artifactType (and chart name/version annotations) on helm push so
  charts are identifiable at the Index level without fetching each
  manifest.
- Add a --subject flag to helm push that associates the chart with an
  existing artifact in the same repository via the OCI Referrers API;
  the digest is validated and resolved against the target repository to
  a full descriptor.

Closes #31582

Assisted-By: Claude <noreply@anthropic.com>
Signed-off-by: Aleksei Sviridkin <f@lex.la>
